src/hangar/diagnostics/integrity.py

```diff
@@ -0,0 +1,212 @@
+import os
+import tempfile
+import warnings
+
+import lmdb
+import numpy as np
+from tqdm import tqdm
+
+from .. import constants as c
+from ..backends import BACKEND_ACCESSOR_MAP, is_local_backend
+from ..context import TxnRegister
+from ..records import commiting, hashmachine, hashs, parsing, queries, heads
+
+
+def _verify_array_integrity(hashenv: lmdb.Environment, repo_path: os.PathLike):
+
+    fs = {}
+    hq = hashs.HashQuery(hashenv)
+    data_kvs = hq.gen_all_hash_keys_raw_array_vals_parsed()
+    narrays = hq.num_arrays()
+    nremote = 0
+    try:
+        for digest, val in tqdm(data_kvs, total=narrays, desc='verifying arrays'):
+            if val.backend not in fs:
+                fs[val.backend] = BACKEND_ACCESSOR_MAP[val.backend](repo_path, None, None)
+                fs[val.backend].open(mode='r')
+            if is_local_backend(val) is False:
+                nremote += 1
+                warnings.warn(
+                    'Can not verify integrity of partially fetched array data references. '
+                    'For complete proof, fetch all remote data locally.', RuntimeWarning)
+                continue
+            o = fs[val.backend].read_data(val)
+            tcode = hashmachine.hash_type_code_from_digest(digest)
+            calc_digest = hashmachine.array_hash_digest(array=o, tcode=tcode)
+            if calc_digest != digest:
+                raise RuntimeError(
+                    f'Data corruption detected for array. Expected digest `{digest}` '
+                    f'currently mapped to spec `{val}`. Found digest `{calc_digest}`')
+        if nremote > 0:
+            warnings.warn(
+                f'Num unverified remote arrays {nremote}/{narrays}.', RuntimeWarning)
+    finally:
+        for be in fs.keys():
+            fs[be].close()
+
+
+def _verify_schema_integrity(hashenv: lmdb.Environment):
+
+    hq = hashs.HashQuery(hashenv)
+    schema_kvs = hq.gen_all_schema_keys_raw_vals_parsed()
+    nschemas = hq.num_schemas()
+    for digest, val in tqdm(schema_kvs, total=nschemas, desc='verifying schemas'):
+        tcode = hashmachine.hash_type_code_from_digest(digest)
+        calc_digest = hashmachine.schema_hash_digest(
+            shape=val.schema_max_shape,
+            size=np.prod(val.schema_max_shape),
+            dtype_num=val.schema_dtype,
+            named_samples=val.schema_is_named,
+            variable_shape=val.schema_is_var,
+            backend_code=val.schema_default_backend,
+            backend_opts=val.schema_default_backend_opts,
+            tcode=tcode)
```

Comment:
This whole section is particularly susceptible to breakage. If the schema value is corrupted or changed, theres no guarantee that the parsed values would contain the expected field names (if it was even able to be deserialized at all)